ast/output_test.go

```diff
@@ -0,0 +1,215 @@
+package ast
+
+import (
+	"testing"
+)
+
+func TestOutput_type(t *testing.T) {
+	testCases := []struct{
+		Name string
+		Output *Output
+		Scope Scope
+		ReturnType Type
+		ShouldError bool
+	}{
+		{
+			Name: "No expressions, for backward compatibility",
+			Output: &Output{},
+			Scope: nil,
+			ReturnType: TypeString,
+		},
+		{
+			Name: "Single string expression",
+			Output: &Output{
+				Exprs: []Node{
+					&LiteralNode{
+						Value: "Whatever",
+						Typex: TypeString,
+					},
+				},
+			},
+			Scope: nil,
+			ReturnType: TypeString,
+		},
+		{
+			Name: "Single list expression of strings",
+			Output: &Output{
+				Exprs: []Node{
+					&VariableAccess{
+						Name: "testvar",
+					},
+				},
+			},
+			Scope: &BasicScope{
+				VarMap: map[string]Variable{
+					"testvar": Variable{
+						Type: TypeList,
+						Value: []Variable{
+							Variable{
+								Type: TypeString,
+								Value: "Hello",
+							},
+							Variable{
+								Type: TypeString,
+								Value: "World",
+							},
+						},
+					},
+				},
+			},
+			ReturnType: TypeList,
+		},
+		{
+			Name: "Single map expression",
+			Output: &Output{
+				Exprs: []Node{
+					&VariableAccess{
+						Name: "testvar",
+					},
+				},
+			},
+			Scope: &BasicScope{
+				VarMap: map[string]Variable{
+					"testvar": Variable{
+						Type: TypeMap,
+						Value: map[string]Variable{
+							"key1": Variable{
+								Type: TypeString,
+								Value: "Hello",
+							},
+							"key2": Variable{
+								Type: TypeString,
+								Value: "World",
+							},
+						},
+					},
+				},
+			},
+			ReturnType: TypeMap,
+		},
+		{
+			Name: "Multiple map expressions",
+			Output: &Output{
+				Exprs: []Node{
+					&VariableAccess{
+						Name: "testvar",
+					},
+					&VariableAccess{
+						Name: "testvar",
+					},
+				},
+			},
```

Comment:
So this is equivalent to `"${var.foo}${var.foo}"` and that's illegal because what would the resulting type even be. Am I reading this right?